File: leetcode/C++/reduce-array-size-to-the-half.cpp
class Solution {
public:
    int minSetSize(vector<int>& arr) {
        int sz = arr.size(), t=arr.size(), sol = 0;
        vector<int> v;
        unordered_map<int,int> um;
        for(int &t:arr) ++um[t];
        for(const pair<int,int> &p : um) v.push_back(p.second);
        sort(v.begin(),v.end(), [](int &a, int &b) {return a>b;});
        for(int i=0; i<v.size(); ++i){
            t-=v[i]; ++sol;
            if(sz/2>=t) break;
        }
        return sol;
    }
};

File: leetcode/Python/reduce-array-size-to-the-half.py
class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        um, v, sol, t = defaultdict(int), [], 0, len(arr)
        for i in arr:
            um[i]+=1
        for i,val in um.items():
            v.append(val)
        v.sort()
        for i in range(len(v)-1,-1,-1):
            t-=v[i]
            sol+=1
            if len(arr)//2 >= t:
                break
        return sol